core/src/replay_stage.rs

```diff
@@ -590,6 +590,13 @@ impl ReplayStage {
                 trace!("bank has_voted: {} {}", b.slot(), has_voted);
                 !has_voted
             })
+            .filter(|b| {
```

Comment:
I think the `frozen_banks`  which is  the subject of the filter here,  is calculated from `BankForks` which should only contain banks > root, so this filter should be redundant

Reply:
Well it's not. May be some timing thing. Should it happen in `prune_non_root`?

Reply:
replay calls `generate_votable_banks` where this logic happens before going into `handle_votable_bank` where the set_root is called.

Reply:
Yeah, so `prune_non_root` should clear all banks < root. Then no new roots will be set until we commit to voting on a bank in `handle_votable_bank`. Because `generate_votable_banks` happens before `handle_votable_bank` this implies BankForks should already have run `prune_non_root` on the latest root, so any filter you apply here should be redundant?

Reply:
`prune_non_root` won't get called until `handle_votable_bank` is called. So the root will not be pruned before this filter.

Reply:
This definitely has an effect. Before, there is a message about the vote being dropped:

```
[2019-09-19T17:55:54.616102076Z WARN  solana_vote_api::vote_state] EoqfC2zWinnd2uN95HJZSghtbAH7NZmYvsMdw8PPG3pg dropped vote Vote { slots: [830], hash: 1111111111111111111111111111111 } too old: [(830, 11111111111111111111111111111111)]
```

Reply:
hmmm I see, it's because it's checking the `is_locked_out` check on the root itself as well, and your check b/c it's a strict > check, removes the root